### PR TITLE
Fix SSE detection on non-AVX CPUs

### LINKED ISSUE #4: SSSE3, SSE4.1, SSE4.2 not detected on non-AVX CPUs
ParseCpuId is coded to require the XMM XCR0 bit in order to recognize SSSE3, SSE4.1, and SSE4.2, but this is not the right way to detect it.  Machines without AVX will not have XCR0 at all, and will fail this check.

In the absence of XCR0, on x86-32, the proper way to detect SSE is to ask the operating system.  (On x86-64, just assume yes.)  In Windows, that's IsProcessorFeaturePresent(PF_XMMI_INSTRUCTIONS_AVAILABLE).  In Linux, unfortunately, it's parsing /proc/cpuinfo.  In macOS and iOS (simulator), it's sysctlbyname on hw.optional.sse, though you could also just assume yes.

Something could be said that the number of x86 machines without SSE in the CPU and OS is about zero now--it depends on how far back you want your library to support.

### PR DESCRIPTION
**Pull request by gchatelet**
Fixes #4 

This is based on #115 with a few modifications:
 - Removed use of `__builtin_cpu_supports` since it relies on `cpuid` and doesn't improve on the current situation,
 - Added detection for all of `sse`, `sse2`, `sse3`, `ssse3`, `sse4_1` and `sse4_2`,
 - Added tests for Atom, Nehalem, and P3 processors,

Thx to @gadoofou87 for providing the original PR.
It also removes the need for #92

### CODE REVIEW

**Review by Mizux — APPROVED** [review]

File: src/cpuinfo_x86.c [hunk 1]
```diff
@@ -25,6 +25,21 @@
 #error "Cannot compile cpuinfo_x86 on a non x86 platform."
 #endif
 
+// The following includes are necessary to provide SSE detections on pre-AVX
+// microarchitectures.
+#if defined(CPU_FEATURES_COMPILER_MSC)
+#include <windows.h>  // IsProcessorFeaturePresent
```
Comment (Mizux) [inline src/cpuinfo_x86.c:31]: I fear that by adding this header, it break __cpuindex line 73 on windows
need to investigate....

**Comment by Mizux** [comment]
otherwise `CpuidX86Test.P3` fails on osx (travis-ci) while it pass on xenia (travis-ci)l

File: src/cpuinfo_x86.c [hunk 2]
```diff
@@ -1082,27 +1097,106 @@ static void ParseLeaf4(const int max_cpuid_leaf, CacheInfo* info) {
...
+  uname(&buf);
+  if (CpuFeatures_StringView_IsEquals(str(buf.sysname), str("Darwin"))) {
+#if defined(HAVE_SYSCTLBYNAME)
+    // Handling Darwin platform through sysctlbyname when available.
+    features->sse = SysCtlByName("hw.optional.sse");
+    features->sse2 = SysCtlByName("hw.optional.sse2");
+    features->sse3 = SysCtlByName("hw.optional.sse3");
+    features->ssse3 = SysCtlByName("hw.optional.supplementalsse3");
+    features->sse4_1 = SysCtlByName("hw.optional.sse4_1");
+    features->sse4_2 = SysCtlByName("hw.optional.sse4_2");
```
Comment (Mizux) [inline src/cpuinfo_x86.c:1158]: so on Apple for fake_cpu you also need to moc SysCtlByName...

**Review by Mizux — CHANGES_REQUESTED** [review]
see comments

File: src/cpuinfo_x86.c (hunk 1 above)
Comment (gchatelet) [reply src/cpuinfo_x86.c:31]: Yep, saw that ...
I'll have a look on a Windows box.

File: src/cpuinfo_x86.c (hunk 2 above)
Comment (gchatelet) [reply src/cpuinfo_x86.c:1158]: Yes I was expecting this :)
Working on a fix.

File: src/cpuinfo_x86.c (hunk 1 above)
Comment (Mizux) [reply src/cpuinfo_x86.c:31]: Mabe replace it by `processthreadsapi.h` ?
ref: https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-isprocessorfeaturepresent

Comment (Mizux) [reply src/cpuinfo_x86.c:31]: not working:
```sh
mizux@DESKTOP-SUSBN78 MINGW64 ~/work/cpu_features (fix_sse_detection_on_pre_avx_machines)
$ cmake --build build --config Debug
Microsoft (R) Build Engine version 16.6.0+5ff7b0c9e for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/Users/mizux/work/cpu_features/CMakeLists.txt
  filesystem.c
  stack_line_reader.c
  string_view.c
  Generating Code...
  utils.vcxproj -> C:\Users\mizux\work\cpu_features\build\utils.dir\Debug\utils.lib
  Building Custom Rule C:/Users/mizux/work/cpu_features/CMakeLists.txt
  cpuinfo_x86.c
C:\Users\mizux\work\cpu_features\src\cpuinfo_x86.c(73,6): warning C4391: 'Leaf __cpuidex(uint32_t,int)': incorrect return type for intrinsic function, expected 'void' [C:\Users\mizux\work\cpu_features\build\cpu_features.vcxproj]
C:\Users\mizux\work\cpu_features\src\cpuinfo_x86.c(73,6): warning C4392: 'Leaf __cpuidex(uint32_t,int)': incorrect number of arguments for intrinsic function, expected '3' arguments [C:\Users\mizux\work\cpu_features\build\cpu_features.vcxproj]
C:\Users\mizux\work\cpu_features\src\cpuinfo_x86.c(73,41): error C2169: '__cpuidex': intrinsic function, cannot be defined [C:\Users\mizux\work\cpu_features\build\cpu_features.vcxproj]
C:\Users\mizux\work\cpu_features\src\cpuinfo_x86.c(90,64): error C2168: '__cpuidex': too few actual parameters for intrinsic function [C:\Users\mizux\work\cpu_features\build\cpu_features.vcxproj]
C:\Users\mizux\work\cpu_features\src\cpuinfo_x86.c(96,32): error C2168: '__cpuidex': too few actual parameters for intrinsic function [C:\Users\mizux\work\cpu_features\build\cpu_features.vcxproj]

mizux@DESKTOP-SUSBN78 MINGW64 ~/work/cpu_features (fix_sse_detection_on_pre_avx_machines)
$ cmake --build build --config Debug
Microsoft (R) Build Engine version 16.6.0+5ff7b0c9e for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  utils.vcxproj -> C:\Users\mizux\work\cpu_features\build\utils.dir\Debug\utils.lib
  cpuinfo_x86.c
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\winnt.h(173,1): fatal error C1189: #error:  "No Target Architecture" [C:\Users\mizux\work\cpu_features\build\cpu_features.vcxproj]
```

Comment (Mizux) [reply src/cpuinfo_x86.c:31]: FYI it seems `__cpuidex` and `CpuIdEx` conflict (same mangle ?) try to rename it and it build
quick and dirty fix rename it CpuIdExWrap....
```cpp
Leaf CpuIdExWrap(uint32_t leaf_id, int ecx) {
  Leaf leaf;
  int data[4];
  __cpuidex(data, (int)leaf_id, ecx);
  leaf.eax = data[0];
  leaf.ebx = data[1];
  leaf.ecx = data[2];
  leaf.edx = data[3];
  return leaf;
}

uint32_t GetXCR0Eax(void) { return (uint32_t)_xgetbv(0); }

#else
#error "Unsupported compiler, x86 cpuid requires either GCC, Clang or MSVC."
#endif

static Leaf CpuId(uint32_t leaf_id) { return CpuIdExWrap(leaf_id, 0); }

static const Leaf kEmptyLeaf;

static Leaf SafeCpuIdEx(uint32_t max_cpuid_leaf, uint32_t leaf_id, int ecx) {
  if (leaf_id <= max_cpuid_leaf) {
    return CpuIdExWrap(leaf_id, ecx);
```

Comment (gchatelet) [reply src/cpuinfo_x86.c:31]: Thx!

**Review by Mizux — APPROVED** [review]

File: test/cpuinfo_x86_test.cc [hunk 3]
```diff
@@ -52,11 +53,22 @@ class FakeCpu {
...
+  bool SetWindowsIsProcessorFeaturePresent(unsigned long ProcessorFeature) {
+    windows_isprocessorfeaturepresent_.insert(ProcessorFeature);
+  }
+#endif  // CPU_FEATURES_OS_WINDOWS
 
  private:
   std::map<std::pair<uint32_t, int>, Leaf> cpuid_leaves_;
-  std::map<std::string, int> sysctlbyname_;
+  std::set<std::string> darwin_sysctlbyname_;
+  std::set<unsigned> windows_isprocessorfeaturepresent_;
```
Comment (Mizux) [inline test/cpuinfo_x86_test.cc:74]: unsigned long ?

File: test/cpuinfo_x86_test.cc [hunk 4]
```diff
@@ -36,12 +37,12 @@ class FakeCpu {
...
-    const auto itr = sysctlbyname_.find(name);
-    if (itr != sysctlbyname_.end()) {
-      return itr->second;
-    }
-    return 0;
+  bool GetDarwinSysCtlByName(std::string name) const {
+    return darwin_sysctlbyname_.count(name);
+  }
+
+  bool GetWindowsIsProcessorFeaturePresent(unsigned long ProcessorFeature) {
```
Comment (Mizux) [inline test/cpuinfo_x86_test.cc:47]: why not using DWORD type here (to be consistent with the test line ~80) ?

Comment (gchatelet) [reply test/cpuinfo_x86_test.cc:47]: Good catch, this was before I included `windows.h` :)

File: test/cpuinfo_x86_test.cc (hunk 3 above)
Comment (gchatelet) [reply test/cpuinfo_x86_test.cc:74]: :+1:

File: test/cpuinfo_x86_test.cc [hunk 5]
```diff
@@ -57,21 +49,33 @@ class FakeCpu {
...
 #endif  // CPU_FEATURES_OS_WINDOWS
 
  private:
   std::map<std::pair<uint32_t, int>, Leaf> cpuid_leaves_;
+#if defined(CPU_FEATURES_OS_DARWIN)
   std::set<std::string> darwin_sysctlbyname_;
-  std::set<unsigned> windows_isprocessorfeaturepresent_;
+#endif  // CPU_FEATURES_OS_DARWIN
+#if defined(CPU_FEATURES_OS_WINDOWS)
+  std::set<unsigned long> windows_isprocessorfeaturepresent_;
```
Comment (Mizux) [inline test/cpuinfo_x86_test.cc:77]: s/unsigned long/DWORD ?

**Review by Mizux — CHANGES_REQUESTED** [review]

**Review by Mizux — APPROVED** [review]

**Comment by gchatelet** [comment]
Thx for the review @Mizux !

**Pull request merged**